components/dashboard/src/settings/Integrations.tsx

```diff
@@ -199,77 +179,435 @@ function GitProviders() {
+
+            // close the modal, as the creation phase is done anyways.
+            close();
+        } catch (error) {
+            console.log(error);
+            setErrorMessage("message" in error ? error.message : "Failed to update Git provider");
+        }
+        setBusy(false);
+    }
+
+    const updateHostValue = (host: string) => {
+        if (props.mode === "new") {
+            setHost(host);
+            setRedirectURL(callbackUrl(host));
+            setErrorMessage(undefined);
+        }
+    }
+
+    const updateClientId = (value: string) => {
+        setClientId(value);
+    }
+    const updateClientSecret = (value: string) => {
+        setClientSecret(value);
+    }
+
+    const validate = () => {
+        const errors: string[] = [];
+        if (clientId.trim().length === 0) {
+            errors.push("Client ID is missing.");
+        }
+        if (clientSecret.trim().length === 0) {
+            errors.push("Client Secret is missing.");
+        }
+        if (errors.length === 0) {
+            setValidationError(undefined);
+            return true;
+        } else {
+            setValidationError(errors.join("\n"));
+            return false;
+        }
+    }
+
+    const getRedirectUrlDescription = (type: string, host: string) => {
+        let url = ``;
+        switch (type) {
+            case "GitHub":
+                url = `${host}/settings/developers`;
+                break;
+            case "GitLab":
+                url = `${host}/profile/applications`;
+                break;
+            default: return undefined;
+        }
+
+        return (<span>
+            Use this redirect URL to update the OAuth application.
+            Go to <a href={`https://${url}`} target="_blank" rel="noopener" className="text-gray-400 underline underline-thickness-thin underline-offset-small hover:text-gray-600">{url}</a> and setup the OAuth application.&nbsp;
+            <a href="https://www.gitpod.io/docs/gitlab-integration/#oauth-application" target="_blank" rel="noopener" className="text-gray-400 underline underline-thickness-thin underline-offset-small hover:text-gray-600">Learn more</a>.
+        </span>);
+    }
+
+    const copyRedirectUrl = () => {
+        const el = document.createElement("textarea");
+        el.value = redirectURL;
+        document.body.appendChild(el);
+        el.select();
+        try {
+            document.execCommand("copy");
+        } finally {
+            document.body.removeChild(el);
+        }
+    };
+
+    return (<Modal visible={!!props} onClose={close}>
+        <h3 className="pb-2">{props.mode === "new" ? "New Git Integration" : "Git Integration"}</h3>
+        <div className="space-y-4 border-t border-b border-gray-200 mt-2 -mx-6 px-6 py-4">
+            {props.mode === "edit" && props.provider.status === "pending" && (
+                <div className="flex rounded-md bg-gitpod-kumquat-light p-3">
```

Comment:
**thought**: It would be nice to include the exclamation mark here before the text as seen in the specs. However, feel free to leave this as is anD create a follow-up issue for this tackle in a later iteration. 🏓